### ISSUE TITLE
Missing type parameter RuntimeException

### ISSUE THREAD

**Issue by usmanrana-ilsa** [google/gson#2069]
# Gson version
'2.8.9'

# Java / Android version
Java 11 / Android all

I'm facing following exception with Gradle version '7.1.0', whereas '7.0.4' is working fine. The issue is happening in both cases when minifyEnabled = true/false in that module, but main app module has minifyEnabled=true and proguard rules are mentioned as well.

```
Fatal Exception: java.lang.RuntimeException
Missing type parameter.
com.google.gson.reflect.TypeToken.getSuperclassTypeParameter (TypeToken.java:84)
com.google.gson.reflect.TypeToken.<init> (TypeToken.java:62)
```

**Comment by Marcono1234** [comment]
Could you please provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example)?

**Comment by usmanrana-ilsa** [comment]
- Put Gson in a library module with release as mindifyEnabled=false

**Comment by Marcono1234** [comment]
Just to verify, your ProGuard rules contain the following, correct?
```
-keepattributes Signature
```

[This StackOverflow question](https://stackoverflow.com/q/70969756) seems to describe the same issue for Gradle 7.1.0, and also contains a (slightly cumbersome and error-prone) workaround. It would it be interesting nonetheless to understand why this started appearing with Gradle 7.1.0.

Unfortunately I am not so familiar with Android development, and don't have the necessary development setup at the moment. Would be great if someone more familiar with Android, or one of the Gson project members could have a look.

**Comment by usmanrana-ilsa** [comment]
yes, adding that to ProGuard rules also didn't work. Issue is when using Android Gradle plugin `7.1.0` . Otherwise same code works fine with '`7.0.4`'. Also the stackOverflow link worked for me which recommends using '`Type listType = TypeToken.getParameterized(List.class, Country.class).getType();`'

**Comment by Marcono1234** [comment]
Have you also tried adding the following, which was added in #1930:
```
# Retain generic signatures of TypeToken and its subclasses with R8 version 3.0 and higher.
-keep,allowobfuscation,allowshrinking class com.google.gson.reflect.TypeToken
-keep,allowobfuscation,allowshrinking class * extends com.google.gson.reflect.TypeToken
```

**Comment by keobrien04** [comment]
So I've tried to unsubscribe to these, and GitHub says I'm not subscribed. How do I get off the email chain?On Feb 6, 2022 9:11 AM, Marcono1234 ***@***.***> wrote:
Just to verify, your ProGuard rules contain the following, correct?
-keepattributes Signature

This StackOverflow question seems to describe the same issue for Gradle 7.1.0, and also contains a (slightly cumbersome and error-prone) workaround. It would it be interesting nonetheless to understand why this started appearing with Gradle 7.1.0.
Unfortunately I am not so familiar with Android development, and don't have the necessary development setup at the moment. Would be great if someone more familiar with Android, or one of the Gson project members could have a look.

**Comment by forlayo** [comment]
In my case was just adding the following to proguard configuration:

```
# Retain generic signatures of TypeToken and its subclasses with R8 version 3.0 and higher.
-keep,allowobfuscation,allowshrinking class com.google.gson.reflect.TypeToken
-keep,allowobfuscation,allowshrinking class * extends com.google.gson.reflect.TypeToken
```

Here you go the full set of options that are needed for Gson -> https://github.com/google/gson/blob/master/examples/android-proguard-example/proguard.cfg

**Comment by iamarjun** [comment]
> In my case was just adding the following to proguard configuration:
> 
> ```
> # Retain generic signatures of TypeToken and its subclasses with R8 version 3.0 and higher.
> -keep,allowobfuscation,allowshrinking class com.google.gson.reflect.TypeToken
> -keep,allowobfuscation,allowshrinking class * extends com.google.gson.reflect.TypeToken
> ```
> 
> Here you go the full set of options that are needed for Gson -> https://github.com/google/gson/blob/master/examples/android-proguard-example/proguard.cfg

is `Proguard.cfg` any different from `proguard-rules.pro`?, will adding this to `proguard-rules.pro` work?

**Comment by forlayo** [comment]
@iamarjun It’s same

**Comment by eamonnmcmanus** [comment]
Thanks to everyone who has commented explaining how this can be fixed with ProGuard configuration. Since this isn't a Gson issue as such, I'm going to close it. The information here will remain valuable to others running into the same problem.

**Comment by JonatanGarces** [comment]
Faced the same issue and this solved the issue:

original code: Type type = new TypeToken<ArrayList<String>>(){}.getType();
new code: Type type = TypeToken.getParameterized(ArrayList.class, String.class).getType();

in the build.grade:
implementation 'com.google.code.gson:gson:2.10.1'
implementation 'androidx.lifecycle:lifecycle-livedata-ktx:2.3.1'
implementation 'androidx.lifecycle:lifecycle-viewmodel-ktx:2.3.1'
